source/navier_stokes_projection/assemble_diffusion_rhs.cc

```diff
@@ -1,13 +1,360 @@
 #include <rotatingMHD/navier_stokes_projection.h>
+
 #include <deal.II/base/work_stream.h>
 #include <deal.II/fe/fe_nothing.h>
 #include <deal.II/grid/filtered_iterator.h>
 
 namespace RMHD
 {
 
+namespace
+{
+
+
+template <int dim>
+void compute_body_force
+(TensorFunction<1, dim>* const  ptr,
+ const std::vector<Point<dim>> &quadrature_points,
+ const std::vector<double>     &gamma,
+ const double                   previous_time,
+ const double                   current_time,
+ const double                   next_time,
+ std::vector<Tensor<1,dim>>    &body_force)
+{
+  AssertDimension(body_force.size(), quadrature_points.size());
+
+  std::vector<Tensor<1,dim>>  old_old_body_force_values(quadrature_points.size());
+  ptr->set_time(previous_time);
+  ptr->value_list(quadrature_points,
+                  old_old_body_force_values);
+
+  std::vector<Tensor<1,dim>>  old_body_force_values(quadrature_points.size());
+  ptr->set_time(current_time);
+  ptr->value_list(quadrature_points,
+                  old_body_force_values);
+
+  std::vector<Tensor<1,dim>>  body_force_values(quadrature_points.size());
+  ptr->set_time(next_time);
+  ptr->value_list(quadrature_points,
+                  body_force_values);
+
+  // Loop over quadrature points
+  for (std::size_t q=0; q<quadrature_points.size(); ++q)
+    body_force[q] =
+      (gamma[0] * body_force_values[q] +
+       gamma[1] * old_body_force_values[q] +
+       gamma[2] * old_old_body_force_values[q]);
+}
```

Comment:
This method could be generalized by taking another template argument (scalar or vector valued) and renaming it to `compute_supply_term`